play/src/front/Chat/Connection/Matrix/MatrixChatConnection.ts

```diff
@@ -121,56 +133,299 @@ export class MatrixChatConnection implements ChatConnectionInterface {
+            });
+    }
+    private createAndAddNewRootRoom(room: Room): MatrixChatRoom {
+        const newRoom = new MatrixChatRoom(room);
+        this.roomList.set(newRoom.id, newRoom);
+        return newRoom;
+    }
     private onClientEventDeleteRoom(roomId: string) {
-        this.roomList.delete(roomId);
+        this.deleteRoom(roomId);
     }
+    private deleteRoom(roomId: string) {
+        const isRootRoom = this.roomList.delete(roomId);
+        if (isRootRoom) {
+            return;
+        }
+        const isRootFolder = this.roomFolders.has(roomId);
 
+        if (isRootFolder) {
+            return;
+        }
+
+        this.roomFolders.forEach((folder) => {
+            folder.deleteNode(roomId);
+        });
+
+        const currentRoom = get(selectedRoom)?.id;
+        if (currentRoom && currentRoom === roomId) selectedRoom.set(undefined);
+    }
     private onRoomEventMembership(room: Room, membership: string, prevMembership: string | undefined) {
         const { roomId } = room;
-        const existingMatrixChatRoom = this.roomList.has(roomId);
+        const existingMatrixChatRoom = this.roomList.has(roomId) || this.roomFolders.has(roomId);
+
         if (membership !== prevMembership && existingMatrixChatRoom) {
             if (membership === KnownMembership.Join) {
-                this.roomList.set(roomId, new MatrixChatRoom(room));
-                return;
-            }
-            if (membership === KnownMembership.Leave || membership === KnownMembership.Ban) {
                 this.roomList.delete(roomId);
-                const currentRoom = get(selectedRoom)?.id;
-                if (currentRoom && currentRoom === roomId) selectedRoom.set(undefined);
+                this.manageRoomOrFolder(room);
                 return;
             }
 
             if (membership === KnownMembership.Invite) {
-                const newRoom = new MatrixChatRoom(room);
-                this.roomList.set(roomId, newRoom);
+                this.createAndAddNewRootRoom(room);
                 return;
             }
         }
-    }
 
+        if (membership === KnownMembership.Leave || membership === KnownMembership.Ban) {
+            this.deleteRoom(roomId);
+            return;
+        }
+    }
     //TODO createOptions only on matrix size
     async createRoom(roomOptions?: CreateRoomOptions): Promise<{
         room_id: string;
     }> {
         if (roomOptions === undefined) {
             return Promise.reject(new Error("CreateRoomOptions is empty"));
         }
+
         try {
-            return await this.client.createRoom(this.mapCreateRoomOptionsToMatrixCreateRoomOptions(roomOptions));
+            const result = await this.client.createRoom(
+                this.mapCreateRoomOptionsToMatrixCreateRoomOptions(roomOptions)
+            );
+
+            await new Promise<void>((resolve, _) => {
+                this.client.once(ClientEvent.Sync, (state) => {
```

Comment:
You use "once" so the event will trigger only once.
What if the state triggered is NOT the "syncing" state? If so, the event handler will be unregistered, and the promise will never resolve.

Reply:
Try putting this logic in a function.

Reply:
Shall we put a timeout here? If the sync does not happen within 20-30 seconds?